UnityProject/Assets/Scripts/UI/Objects/Medical/genetics/SudokuPuzzle/sudoku generator/SudokuGenerator.cs

```diff
@@ -0,0 +1,600 @@
+            blank_board.Add('.');
+        }
+        var candidates = _get_candidates_map(blank_board);
+
+        // For each item in a shuffled list of squares
+        var shuffled_squares = _shuffle(SQUARES);
+        foreach(var si in shuffled_squares){
+            var square = si;
+
+            // If an assignment of a random chioce causes a contradictoin, give
+            // up and try again
+            var rand_candidate_idx = _rand_range(candidates[square].Length);
+            var rand_candidate = candidates[square][rand_candidate_idx];
+            if(_assign(candidates, square, rand_candidate) == null){
+                break;
+            }
+
+            // Make a list of all single candidates
+            var single_candidates = new List<string>();
+            foreach(var sAi in SQUARES){
+                var square2 = sAi;
+
+                if(candidates[square2].Length == 1){
+                    single_candidates.Add(candidates[square2]);
+                }
+            }
+
+            // If we have at least difficulty, and the unique candidate count is
+            // at least 8, return the puzzle!
+            if(single_candidates.Count >= difficulty && _strip_dups(single_candidates).Count >= 8){
+                var board = "";
+                var givens_idxs = new List<int>();
+                for (int i = 0; i < SQUARES.Count; i++)
+                {
+	                var square2 = SQUARES[i];
+                    if(candidates[square2].Length == 1){
+                        board += candidates[square2];
+                        givens_idxs.Add(i);
+                    } else {
+                        board += BLANK_CHAR;
+                    }
+                }
+
+                // If we have more than `difficulty` givens, remove some random
+                // givens until we're down to exactly `difficulty`
+                var nr_givens = givens_idxs.Count;
+                if(nr_givens > difficulty){
+	                givens_idxs = givens_idxs.Shuffle().ToList();
+                    for(var i = 0; i < nr_givens - difficulty; ++i){
+                        var target = givens_idxs[i];
+                        board = board.Substring(0, target) + BLANK_CHAR +
+                            board.Substring(target + 1);
+                    }
+                }
+
+                // Double check board is solvable
+                // TODO: Make a standalone board checker. Solve is expensive.
+                if( string.IsNullOrEmpty(solve(board)) == false){
+                    return board;
+                }
+            }
+        }
+
+        // Give up and try a new puzzle
+        return generate(InDIFFICULTY);
+    }
+
+
+	public Dictionary<string, string> _get_candidates_map(List<char> board){
+		/* Get all possible candidates for each square as a map in the form
+		{square: sudoku.DIGITS} using recursive constraint propagation. Return `false`
+		if a contradiction is encountered
```

Comment:
would be nice if we got deleted explanation like this everywhere :)
even better, some of these are best written in xml format above functions so we can read them in IDEs when we hover over them